### PR TITLE
Do not remove container if underlying resource cleanup failed

### PR DESCRIPTION
**Pull request by rhvgoyal**
This should fix issue #10955 and #6746 

This patch series does two things. First of all it does not remove container from a user's view if container removal failed. Secondly, it extends force option (-f) to forcibly remove the container even if resource cleanup failed. That way those user who are willing to let resources leak, can use --force.

### CODE REVIEW

**Comment by rhvgoyal** [comment]
Hmm..., there are more calls to daemon.Rm(). Will fix those to adjust for additional parameter.

**Comment by rhvgoyal** [comment]
Pushed one more time fixing all the instances of daemon.Rm(). Hopefully I got them all.

**Comment by jfrazelle** [comment]
can you add fixes #6746 to your first comment so when this is merged it will close that issue as well

**Comment by jfrazelle** [comment]
Also is there anyway to add an integration test

**Comment by rhvgoyal** [comment]
@jfrazelle 

Where should I add "fixes #6746" now? Create a new pull request?

**Comment by rhvgoyal** [comment]
@jfrazelle 

I just edited my initial pull request comment. I guess that's what you were referring to.

I will look into current integration tests and see if they are sufficient to test this or not.

**Comment by jfrazelle** [comment]
oh just in the first comment

**Comment by jfrazelle** [comment]
editing the comment should do it, no need for a new PR

**Comment by rhvgoyal** [comment]
@jfrazelle 

I see that integration-cli/docker_cli_rm_test.go already has bunch of container removal tests. I think adding two more tests to it should make sense.

- Create a failure situation so that container removal fails and container state is "Dead".
- Use "docker rm -f" to be able to remove a dead container.

But I am not sure how to create that failure situation reliably. In my testing I use device mapper driver and bind mount container thing device somewhere so that device deactivation fails which in turn leads to container removal failure. But it is not generic method and I can't see an easy way to force it. 

I am looking for something which can be easily used with all graphdrivers and can cause container removal failure. Any ideas?

**Comment by cpuguy83** [comment]
@rhvgoyal Really you should be able to mount anything into the container fs path and then try to remove the container, this should result in `Device or resource busy` shouldn't it, since there is another fs mounted on top?

**Comment by rhvgoyal** [comment]
@cpuguy83 

I think there are couple of issues.

- I am assuming that daemon is running in a separate mount namespace. And if I bind mount something in client namespace, that would not help.

- Even if I figure out a way to get into daemon mount namespace and bind mount something else on container rootfs, it will not work with my pending proposed patches.  I have proposed another set of patches to unmount container rootfs from daemon mount namespace after successful container start to solve various issues. Once that is in, that means,  container rootfs has already been unmounted and once container exits only device will be deactivated. So bind mounting something on container rootfs will not help.

I proposed my patches here.

https://github.com/docker/docker/pull/11114

**Comment by rhvgoyal** [comment]
@jfrazelle 

So generic container removal tests are already there. What is missing is couple of tests for testing behavior in case of failure. And right now I can't find a good way to force failure.

Can these patches be merged without tests.

**Comment by rhatdan** [comment]
@jfrazelle @crosbymichael @cpuguy83 We would like to get this in for docker-1.6 if possible.

**Comment by jfrazelle** [comment]
@rhatdan I added the milestone

File: daemon/delete.go [hunk 1]
```diff
@@ -61,8 +61,13 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
...
 			}
 		}
-		if err := daemon.Rm(container); err != nil {
-			return job.Errorf("Cannot destroy container %s: %s", name, err)
+		if err := daemon.Rm(container, forceRemove); err != nil {
+			if forceRemove {
+				log.Errorf("Cannot destroy container %s: %s", name, err)
+			} else {
+
+				return job.Errorf("Cannot destroy container %s: %s", name, err)
```
Comment (LK4D4) [inline daemon/delete.go]: This condition can be reversed for readability:
```
if !forceRemove {
    return job.Errorf("Cannot destroy container %s: %s", name, err)
}
log.Errorf("Cannot destroy container %s: %s", name, err)
```

File: daemon/delete.go [hunk 2]
```diff
@@ -92,13 +97,29 @@ func (daemon *Daemon) Rm(container *Container) error {
...
+
 	if err := container.Stop(3); err != nil {
 		return err
 	}
 
-	// Deregister the container before removing its directory, to avoid race conditions
-	daemon.idIndex.Delete(container.ID)
-	daemon.containers.Delete(container.ID)
+	// Mark container dead. We don't want anybody to be restarting it.
+	container.SetDead()
```
Comment (LK4D4) [inline daemon/delete.go]: Shouldn't you check error here?

File: daemon/state.go [hunk 3]
```diff
@@ -217,3 +232,26 @@ func (s *State) IsPaused() bool {
...
+	return nil
+}
+
+func (s *State) ResetRemovalInProgress() {
+	s.Lock()
+	s.RemovalInProgress = false
+	s.Unlock()
+}
+
+func (s *State) SetDead() error {
```
Comment (LK4D4) [inline daemon/state.go]: Ah, so, there is no returning error.

File: daemon/delete.go (hunk 1 above)
Comment (rhvgoyal) [inline daemon/delete.go]: Ok, Will change. This does look more readable.

File: daemon/delete.go (hunk 2 above)
Comment (rhvgoyal) [inline daemon/delete.go]: As you noted in next comment, SetDead() does not return error.

**Comment by rhvgoyal** [comment]
Pushed top most patch again to take care of comment from @LK4D4

File: daemon/container.go [hunk 4]
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
 		return nil
 	}
 
+	if container.RemovalInProgress {
+		return fmt.Errorf("Container is already being removed. Can not start it.")
```
Comment (cpuguy83) [inline daemon/container.go]: Can remove `already`, which is confusing because the user is calling `start`

File: daemon/container.go [hunk 5]
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
...
+	}
+
+	if container.Dead {
+		return fmt.Errorf("Container is already dead. Can not start it.")
```
Comment (cpuguy83) [inline daemon/container.go]: Can remove already, which is confusing because the user is calling start
Also, what does `dead` mean?

File: daemon/delete.go [hunk 6]
```diff
@@ -92,13 +95,29 @@ func (daemon *Daemon) Rm(container *Container) error {
...
+
+	// If force removal is required, delete container from various
+	// indexes even if removal failed.
+	defer func() {
+		if err != nil && forceRemove {
```
Comment (cpuguy83) [inline daemon/delete.go]: I think this `err` is not in scope of any of the bellow `err`'s (or any of the `err's in the whole function), so would always be nil.

File: daemon/state.go (hunk 3 above)
Comment (cpuguy83) [inline daemon/state.go]: If this is not returning an error, can we remove `error` here?

File: daemon/container.go [hunk 7]
```diff
@@ -326,6 +326,14 @@ func (container *Container) Start() (err error) {
 		return nil
 	}
 
+	if container.RemovalInProgress {
```
Comment (cpuguy83) [inline daemon/container.go]: We can probably condense these two errors into one because in terms of the end user, once a container is marked for removal at all it can't be used anyway (regardless if it got to `dead` state or not).

So:
```
if container.RemovalInProgress || container.Dead {
    return fmt.Errorf("container is marked for removal and cannot be started")
}
```

File: daemon/delete.go [hunk 8]
```diff
@@ -61,8 +61,11 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
 				return job.Errorf("Conflict, You cannot remove a running container. Stop the container before attempting removal or use -f")
 			}
 		}
-		if err := daemon.Rm(container); err != nil {
-			return job.Errorf("Cannot destroy container %s: %s", name, err)
+		if err := daemon.Rm(container, forceRemove); err != nil {
+			if !forceRemove {
+				return job.Errorf("Cannot destroy container %s: %s", name, err)
+			}
+			log.Errorf("Cannot destroy container %s: %s", name, err)
```
Comment (cpuguy83) [inline daemon/delete.go]: Should be `"Cannot destroy container %s: %v"

File: daemon/container.go (hunk 4 above)
Comment (rhvgoyal) [inline daemon/container.go]: Makes sense. Will do.

File: daemon/container.go (hunk 5 above)
Comment (rhvgoyal) [inline daemon/container.go]: Dead means that container can't be started now. One can only try to remove it again. (After previous removal attempt had failed). A dead container might have partially cleaned up its resources already and only some of the resources still might be there.

For example, If I bind mount a container device somewhere, then container removal will fail. It will be in "Removal In Progress" state for 10 seconds and then transition into "Dead" state. 

A user can try to figure out what was wrong, why container removal failed. Fix that condition and try removal again. 

Had to keep a separate state to mark that container is partially destroyed otherwise one might want to start it.

File: daemon/delete.go (hunk 8 above)
Comment (rhvgoyal) [inline daemon/delete.go]: Will change.

File: daemon/container.go (hunk 7 above)
Comment (rhvgoyal) [inline daemon/container.go]: Ok, Will do.

File: daemon/container.go (hunk 5 above)
Comment (rhatdan) [inline daemon/container.go]: Would zombie clear it up  :^)

File: daemon/delete.go (hunk 6 above)
Comment (rhvgoyal) [inline daemon/delete.go]: I tested it and err is not nil.

I have defined a named return error as follows.

func (daemon *Daemon) Rm(container *Container, forceRemove bool) (err error)

So this err inside defer must be referring to that. And while I am returning, I am always explicitly returning an error object. So "err" ends up referring to that. 

But if it is confusing and if you have a preferred way to handle it I can do that. I guess I can stop redeclaring err in various calls and that might help.

File: daemon/container.go (hunk 5 above)
Comment (rhvgoyal) [inline daemon/container.go]: @rhatdan 

Zombie sounds good too. I will use whatever people prefer. I personally am little inclined towards "Dead".

**Comment by rhvgoyal** [comment]
Something interesting is happening. The container I removed using -f, comes back to life if I restart the docker service. Wow, I need to look into it. May be we are not deleting underlying files somewhere and restart scans for these files and gets it back.

**Comment by rhvgoyal** [comment]
Ok, /var/lib/docker/container/<container_id> directory is removed after container root and container init root has been removed. 

So it depends on when failure occurs. If failure occurs before container file removal, it will come back to life. And I think it is a bad idea. Container might be half deleted and if somebody tries to start it, bad things will happen. I will save the container state "Dead" back to disk before proceeding with actual deletion. This will make sure that once container is marked dead, it remains dead even if docker is restarted.

**Comment by rhvgoyal** [comment]
Ok, pushed one more time. Have taken care of the review comments so far. And also took care of making "Dead" state persistent to disk.

**Comment by rhvgoyal** [comment]
@jfrazelle 

This windows test failure seems unrelated to my patches.

**Comment by jfrazelle** [comment]
I reran and it passes, sorry we are working through some windows CI things

**Comment by estesp** [comment]
So this is shaping up to be a slightly different implementation of my PR (#10099) from awhile ago; although there are some better choices here about ordering of operations (rather than remove and re-add).  The naming piece is still difficult, but I'm fine with "Dead" if that is reasonable for others.

The one piece I don't like is that I was preserving the failing error--whatever kept the container from being removed successfully is now a transient error that after hours (or days or weeks?) is now lost, although I guess another try at `docker rm` will resurface it?  However, having it available for `docker ps` was something I liked about my impl, although it added the painful UI problem of "way too long/wide output".

File: daemon/delete.go (hunk 6 above)
Comment (cpuguy83) [inline daemon/delete.go]: But the err in defer is nil, see example:
http://play.golang.org/p/rPGaYb3bVA

Unless I'm missing something (which is entirely possible).

Comment (rhvgoyal) [inline daemon/delete.go]: @cpuguy83 

Argument to defer are calculated at the time of defer and that's why you are seeing err nil. In my case, I don't pass err as argument to deferred function.

Comment (rhvgoyal) [inline daemon/delete.go]: For example, I modified your code a bit to not pass err as an argument to deferred function.

https://play.golang.org/p/r9nxzUbbT4

**Comment by rhvgoyal** [comment]
@estesp 

I think storing the error in container is not a bad idea. And it is not orthogonal to my patches. Right now I am trying to fix the problem of not removing container from main view if deletion failed and create new container state to represent that.

If this works well, then next step can be to also store the failure message. And there can be a separate patch for that. But I don't think that storing error should be a requirement for this patch series to go in.

File: daemon/delete.go (hunk 6 above)
Comment (cpuguy83) [inline daemon/delete.go]: Oh, of course! Thanks for taking the time to clarify, sorry for the noise here!

**Comment by jfrazelle** [comment]
look like needs a rebase sorry :(

File: daemon/state.go [hunk 9]
```diff
@@ -217,3 +232,26 @@ func (s *State) IsPaused() bool {
...
+	return nil
+}
+
+func (s *State) ResetRemovalInProgress() {
+	s.Lock()
+	s.removalInProgress = false
+	s.Unlock()
+}
+
+func (s *State) SetDead() error {
```
Comment (LK4D4) [inline daemon/state.go]: It can't return non-nil error, why have confusing signature?

File: daemon/state.go [hunk 10]
```diff
@@ -42,6 +44,14 @@ func (s *State) String() string {
 		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(s.StartedAt)))
 	}
 
+	if s.removalInProgress {
+		return "Removal In Progress"
+	}
+
+	if s.Dead {
```
Comment (LK4D4) [inline daemon/state.go]: I'm not totally sure, but you checking for `s.Dead` twice.

File: daemon/state.go [hunk 11]
```diff
@@ -11,16 +11,18 @@ import (
...
-	ExitCode   int
-	Error      string // contains last known error when starting the container
-	StartedAt  time.Time
-	FinishedAt time.Time
-	waitChan   chan struct{}
+	Running           bool
+	Paused            bool
+	Restarting        bool
+	OOMKilled         bool
+	removalInProgress bool // Not need for this to be persistent on disk.
```
Comment (LK4D4) [inline daemon/state.go]: could you move this below? to have sorta separation between public and private

File: daemon/delete.go [hunk 12]
```diff
@@ -92,19 +95,40 @@ func (daemon *Daemon) Rm(container *Container) error {
...
 
-	if err := container.Stop(3); err != nil {
+	// Container state RemovalInProgress should be used to avoid races.
+	if err = container.SetRemovalInProgress(); err != nil {
+		return fmt.Errorf("Failed to set container state to RemovalInProgress: %s", err)
+	}
+
+	defer container.ResetRemovalInProgress()
+
+	if err = container.Stop(3); err != nil {
```
Comment (LK4D4) [inline daemon/delete.go]: From where `3` number? Shouldn't be that constant?

File: daemon/delete.go [hunk 13]
```diff
@@ -92,19 +95,40 @@ func (daemon *Daemon) Rm(container *Container) error {
...
+			daemon.containers.Delete(container.ID)
+		}
+	}()
+
 	container.derefVolumes()
 	if _, err := daemon.containerGraph.Purge(container.ID); err != nil {
 		log.Debugf("Unable to remove container from link graph: %s", err)
 	}
 
-	if err := daemon.driver.Remove(container.ID); err != nil {
```
Comment (LK4D4) [inline daemon/delete.go]: I don't understand why you need to remove `:`. It is enough if returned value override top-level error.

File: daemon/delete.go [hunk 14]
```diff
@@ -82,7 +85,7 @@ func (daemon *Daemon) DeleteVolumes(volumeIDs map[string]struct{}) {
 }
 
 // Destroy unregisters a container from the daemon and cleanly removes its contents from the filesystem.
-func (daemon *Daemon) Rm(container *Container) error {
+func (daemon *Daemon) Rm(container *Container, forceRemove bool) (err error) {
```
Comment (LK4D4) [inline daemon/delete.go]: Okay, I'm not in favor of this signature. How about helper function and two separate functions `Rm` and `ForceRm` which will call helper with `true` or `false`?

File: daemon/state.go (hunk 9 above)
Comment (rhvgoyal) [inline daemon/state.go]: Makes sense. Will remove it.

File: daemon/state.go (hunk 10 above)
Comment (rhvgoyal) [inline daemon/state.go]: Hm...., I don't understand this comment. I am first checking for Removal In Progress and dispalying that. Yes a container could be dead as well as removal is in progress. In that case I think it makes sense to display "Removal In Progress".

If that's not set, then I check for "Dead" state and display "dead". So where is the redundant check?

File: daemon/state.go (hunk 11 above)
Comment (rhvgoyal) [inline daemon/state.go]: Ok, will do.

File: daemon/delete.go (hunk 12 above)
Comment (rhvgoyal) [inline daemon/delete.go]: That's existing code. I did not introduce it. I agree that using a constant is much better. But that cleanup is probably should be part of a separate patch series.

File: daemon/delete.go (hunk 13 above)
Comment (rhvgoyal) [inline daemon/delete.go]: Hmm..., What's the advantage of retaining ":"?. Only case where we don't want to override err is when we don't want to return actual error. Something like as follows. And there are easy ways to override that in return statement. 

if  err := foo(); err != nil {
     // This is not catastropic. Return nil.
     return
}

Instead following is easier to read because it does not rely on the previous value of err.

if err = foo(); err != nil {
      // This is not a catastropic error
       return nil
}

So not using ":" looks fine to me in above code example.

File: daemon/delete.go (hunk 14 above)
Comment (rhvgoyal) [inline daemon/delete.go]: Ok, I can do that. That will reduce the size of changes as all the existing calls to Rm() will not have to be modified.

**Comment by rhvgoyal** [comment]
@jfrazelle 

Why are we removing this from 1.6.0 milestone?

**Comment by jfrazelle** [comment]
Since its a bug fix we can include it at anytime in an RC but I don't want
it to hold up RC1 unless you can make these changes in the next hour :)

**Comment by rhvgoyal** [comment]
@jfrazelle 

I can do it but I don't like to rush in the changes as it always increases the possibility of breaking something.

All the comments are minor code cleanup commands, nothing major. It does not sound fair that just before RC1, one can give code cleanup comments and block the inclusion of fixes. 

Can't we take these in 1.6.0 and I can open a separate issue to follow up on code cleanup comments.

**Comment by jfrazelle** [comment]
That's sounds reasonable to me

**Comment by rhvgoyal** [comment]
@LK4D4 

Would you have an objection to including these fixes as it is in 1.6.0 and I follow up with your cleanup comments in a separate issue for next release?

File: daemon/state.go (hunk 10 above)
Comment (LK4D4) [inline daemon/state.go]: line 74 I mean

File: daemon/delete.go (hunk 12 above)
Comment (LK4D4) [inline daemon/delete.go]: Ok, agree

File: daemon/state.go (hunk 10 above)
Comment (rhvgoyal) [inline daemon/state.go]: But that's a separate call in separate function. First call is in String() and second call is in StateString().  So does not look like there is redundancy.

Comment (LK4D4) [inline daemon/state.go]: Lol, indeed. Sorry :blush:

File: daemon/delete.go (hunk 13 above)
Comment (LK4D4) [inline daemon/delete.go]: Just to reduce patch size. I'm not insisting if you like it.

**Comment by LK4D4** [comment]
@rhvgoyal Dunno, do you want me to prepare PR for your branch. Because I don't want to have changed signature in release. This can lead to some nasty merges.

**Comment by jfrazelle** [comment]
we can always include it in another RC when it is ready

**Comment by jfrazelle** [comment]
I would rather wait and have it be good

**Comment by rhvgoyal** [comment]
@jfrazelle @LK4D4 

Ok, I will make changes and generate another PR. Hopefully we can pull it in next RC release.

**Comment by jfrazelle** [comment]
well just push to this one :) we can then include all of it in one RC

**Comment by rhvgoyal** [comment]
@jfrazelle 

Ok.

File: daemon/state.go (hunk 11 above)
Comment (rhvgoyal) [inline daemon/state.go]: @LK4D4 

Now I am thinking that it is better to keep it where it is. Reason being that right now all the possible container states are together. If I move it below waitChan, then a single state is out of the group. And I think it is better to keep all states together instead of forcing grouping based on public/private.

**Comment by rhvgoyal** [comment]
@jfrazelle @LK4D4 

Pushed changes again after taking care of some of the comments. Please have another look.

And this build failure seems unrelated to my changes.

**Comment by LK4D4** [comment]
I don't see any big problems
LGTM

**Comment by estesp** [comment]
This looks ready, but I still think it makes sense to test this functionality as it has new codepaths that current tests won't exercise.  Given our `integration-cli` environment is well-known, and many tests assume client == daemon/host, then wouldn't a simple file open inside the container's mount dir disallow unmounting, causing a "busy" error and removal to fail (in the same way that an extra mount would, but much more simply).  The path to the mount point should be easy to discern, given our knowlege of the test harness/daemon config, right?

**Comment by rhvgoyal** [comment]
@estesp 

- How do we know the mount point of container? Docker daemon does  not export that and hardcoding that in tests is not a good idea as if we change it later, test will be broken.

- Mount point will also depend on graphdriver being used. I am not sure how to query daemon for graphdriver in use.

- I did not understand your comment about client == daemon/host. Can you elaborate  bit more. And also point me to a test which assumes it.

**Comment by rhvgoyal** [comment]
@estesp 

Also I think it is not a good idea to assume that container root is mounted in daemon namespace after container launch. I have another PR request waiting to change that behavior.

https://github.com/docker/docker/pull/11114

I think to make it work, we will have to have some sort of mount/unmount commands from docker so that it can mount a container's root for us explicitly and allow us to query mount point of container explicitly. That will work only if daemon is not running in a separate mount namespace.

**Comment by jfrazelle** [comment]
needs a rebase also can you squash to 1 commit otherwise cherry-picking into the release branch is going to be kinda annoying

**Comment by rhvgoyal** [comment]
@jfrazelle 

rebased and put everything into a single patch and pushed one more time. Tests are passing on my local system. So there might be something wrong with the build.

**Comment by cpuguy83** [comment]
Testing this manually and it works.
Looks like CI is having an issue pulling lxc :(

**Comment by jfrazelle** [comment]
If it fails lxc we aren't merging can you rebuild

**Comment by vbatts** [comment]
linuxcontainers.org is down :-(

**Comment by jfrazelle** [comment]
we will wait...

**Comment by vbatts** [comment]
```
--- FAIL: TestEventsFilterContainerID (0.94s)
	docker_cli_events_test.go:355: Expected 3 events, got 2: [2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) create 2015-03-31T17:36:02.000000000Z ab26eddbd0f26fad7fb1263785e701e23d5f5021afbb196a94d4c31c06a97de9: (from busybox:latest) start]
```

**Comment by jfrazelle** [comment]
ya that one has been racy lately

**Comment by jfrazelle** [comment]
@estesp what are your thoughts

**Comment by jfrazelle** [comment]
are you ok with this getting in 1.6.0, it needs to be merged today otherwise, since it is not really a regression and was a bug in earlier versions I will remove from the milestone

**Comment by jfrazelle** [comment]
passes all tests @vbatts you think ok to merge?

**Comment by vbatts** [comment]
so after learning up on this PR, it will make clean up of containers much nicer.
Like happens occasionally where a container fails to be removed (outside mounts, mount namespace overlap, etc), it will leave the container and underlying state to where it needs to be cleaned up by hand.

With this PR, now it is possible to get the error on `docker rm ...` earlier, fix the issue, then `docker rm ...` again. Without having to bring down the daemon for container cleanup.

I've just walked through a couple of cases by hand, and it's improved the experience.

LGTM

File: daemon/delete.go [hunk 15]
```diff
@@ -94,19 +109,40 @@ func (daemon *Daemon) Rm(container *Container) error {
...
+			daemon.containers.Delete(container.ID)
+		}
+	}()
+
 	container.derefVolumes()
 	if _, err := daemon.containerGraph.Purge(container.ID); err != nil {
 		logrus.Debugf("Unable to remove container from link graph: %s", err)
 	}
 
-	if err := daemon.driver.Remove(container.ID); err != nil {
```
Comment (tiborvass) [inline daemon/delete.go]: @rhvgoyal why remove the `:` ?

Comment (tiborvass) [inline daemon/delete.go]: oh for the defer...

File: daemon/delete.go (hunk 6 above)
Comment (tiborvass) [inline daemon/delete.go]: @rhvgoyal shouldn't the condition be `if err == nil || forceRm` ? that way you don't need to copy this part at the end

File: daemon/delete.go [hunk 16]
```diff
@@ -83,8 +90,16 @@ func (daemon *Daemon) DeleteVolumes(volumeIDs map[string]struct{}) {
...
+	return daemon.commonRm(container, false)
+}
+
+func (daemon *Daemon) ForceRm(container *Container) (err error) {
+	return daemon.commonRm(container, true)
+}
+
 // Destroy unregisters a container from the daemon and cleanly removes its contents from the filesystem.
-func (daemon *Daemon) Rm(container *Container) error {
+func (daemon *Daemon) commonRm(container *Container, forceRemove bool) (err error) {
```
Comment (tiborvass) [inline daemon/delete.go]: I'd much rather keep `Rm` as the name, and just add the bool as a param. That way the calling code above can be reduced to: `if err := daemon.Rm(container, forceRemove); err != nil {`

**Comment by jfrazelle** [comment]
ah shoot sorry @tiborvass my GH page hadn't refreshed until after the button click!

File: daemon/delete.go (hunk 16 above)
Comment (rhvgoyal) [inline daemon/delete.go]: @tiborvass 

That's how it was initially and later @LK4D4 preferred it this way. So I changed it.

Comment (rhvgoyal) [inline daemon/delete.go]: For the record, I liked passing force as a parameter to Rm() better instead of creating another variant. It ended up in more lines of changes as every existing call to Rm() had to be modified but I still had liked it better.

Comment (tiborvass) [inline daemon/delete.go]: @rhvgoyal huh I see, I'll try to convince @LK4D4 in another PR :P
